sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java

```diff
@@ -213,6 +219,78 @@ public static ResourceId convertToFileResourceIfPossible(String outputPrefix) {
   private final WritableByteChannelFactory writableByteChannelFactory;
 
   /**
+   * A class that allows value-dependent writes in {@link FileBasedSink}.
+   *
+   * <p>Users can define a custom type to represent destinations, and provide a mapping to turn
+   * this destination type into an instance of {@link FilenamePolicy}.
+   */
+  public abstract static class DynamicDestinations<T, DestinationT>
+      implements HasDisplayData, Serializable {
+    /**
+     * Returns an object that represents at a high level the destination being written to. May not
+     * return null.
+     */
+    public abstract DestinationT getDestination(T element);
+
+    /**
+     * Returns the default destination. This is used for collections that have no elements as the
+     * destination to write empty files to.
+     */
+    public abstract DestinationT getDefaultDestination();
+
+    /**
+     * Returns the coder for {@link DestinationT}. If this is not overridden, then
+     * the coder registry will be use to find a suitable coder. This must be a
+     * deterministic coder, as {@link DestinationT} will be used as a key type in a
+     * {@link org.apache.beam.sdk.transforms.GroupByKey}.
+     */
+    @Nullable
+    public Coder<DestinationT> getDestinationCoder() {
+      return null;
+    }
+
+    /**
+     * Converts a destination into a {@link FilenamePolicy}. May not return null.
+     */
+    public abstract FilenamePolicy getFilenamePolicy(DestinationT destination);
+
+    /**
+    * Populates the display data.
+     */
+    @Override
+    public void populateDisplayData(DisplayData.Builder builder) {
+    }
+
+    // Gets the destination coder. If the user does not provide one, try to find one in the coder
+    // registry. If no coder can be found, throws CannotProvideCoderException.
+    final Coder<DestinationT> getDestinationCoderWithDefault(CoderRegistry registry)
```

Comment:
This is duplicating code in bigquery DynamicDestinations and in KafkaIO - suppose time to introduce a utility, e.g. `Class<?> getTypeParameterOfInterface(Class<?> clazz, Class<?> iface, int index)` in `ReflectHelpers`.

By the way, see https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/FlatMapElements.java#L181